### ISSUE TITLE
[BUG] Killaura ai is borked

### ISSUE THREAD

**Issue by TimofeyFilkin-alt-i-lost-my-codes** [CCBlueX/LiquidBounce#8470]
### Pre-submission checklist

- [x] I have searched through existing issues and confirmed this bug has not been reported yet.
- [x] I have reproduced this on the latest development build.

### LiquidBounce Branch

Nextgen

### LiquidBounce Build/Version

0.38.1 nextgen

### Operating System

Linux

### Minecraft Version

26.1.2 any protocol

### Describe the bug

Killaura ai aura is broken, wrong data is sent into ai, and as a result, without corrections it cant do anything (for me its easier to play legit than with ai killaura, thats how bad it is)
Its backrotate will wobble for long, it will spin around with no reason to do so and such

### Steps to reproduce

Set killaura rotation to ai
Set correction to none
It sucks

### Client Log

```shell
not relevant
```

**Comment by MukjepScarlet** [comment]
@1zun4

**Comment by sprow-dev** [comment]
Please review the code before making strong assumptions like "wrong data is sent into ai". If I remember correctly, you _do_ need to train the AI. If you'd like to confirm that for me, then I'd ask that you close the issue if I am right or let me know if I am wrong. That way, we can get to the bottom of this issue. Thanks!

**Comment by TimofeyFilkin-alt-i-lost-my-codes** [comment]
> If I remember correctly, you _do_ need to train the AI
and its training code is incorrect so it just does some bullshit. The entire system needs to be rebuilt because izuna just blindly accepted some code without looking at it and as a result it was broken for alot of time now

**Comment by TimofeyFilkin-alt-i-lost-my-codes** [comment]
P.S. pytorch is used incorrectly, thats the root problem

**Comment by sprow-dev** [comment]
I was going to ask for more information at first, but then I realized that I'd be unhelpful if I did that so I've decided to get to the root-cause for you.

The project does not appear to make direct use of PyTorch as there are no Python files or Python directories. However, I wanted to see what was actually in place and found Amazon DJL was the backend, which supports PyTorch and TensorFlow replacement APIs that act as Java drop-ins for normally Python-only code. You were almost correct there.

I was initially very wrong and thought that you had to provide a model file yourself. However, I dug deeper into LiquidBounce's DeepLearning code. Here's the real issue, and it's not your fault. It actually *does* have training support and it uses PyTorch files and setups, just not the library itself. Then, in the backend, it uses TensorFlow to train and manage the model. DJL *does* support both, but it **does not** support cross-compatibility between the 2 libraries. Therefore, the main issue is the cross-fire between TensorFlow and PyTorch.

In addition, there are a few more issues that are not worth mentioning as much.
1. It does not normalize values, causing gradient explosions (if the main issue does get fixed).
1. I did also notice a memory leak that will leak a large amount of memory every time you initialize training.

Additionally, I would appreciate if you condense your responses down into one message and checked your markdown formatting before hitting send. I would also prefer a response that does not just use vague derogatory words in place of what is considered "helpful wording". That just helps me and others read what you are saying more easily and helps provide more information. An unwritten rule of open-source is that you should treat issues and pull-requests as if it is the developer's time, not yours. That's just for future consideration and it is simply a mere formality, not a hard requirement. Either way, it will definitely make people unlike me a lot more willing to put time into your issues if you follow that simple rule.

Unfortunately, I do not have the setup or environment to test and update the code and make a PR, but I may be able to let a maintainer or contributor know so that they can fix it up and make it work properly. Does that sound good?

Edit: TL;DR: The root cause is a cross-compatibility conflict within Amazon DJL, which is trying to use PyTorch configurations on a TensorFlow backend, among many other issues. Do you want me to pass this along to a maintainer?

**Comment by TimofeyFilkin-alt-i-lost-my-codes** [comment]
> Do you want me to pass this along to a maintainer?

yeah would be nice if ai killaura would at least work without corrections that flag you

**Comment by 1zun4** [comment]
There is no way I am now being contacted by AI agents with AI slop. Please do not do this.

**Comment by MukjepScarlet** [comment]
> P.S. pytorch is used incorrectly, thats the root problem

Can you explain?

**Comment by TimofeyFilkin-alt-i-lost-my-codes** [comment]
i dont really remember, but there was some shit with data mismatch in
additon to lackage of normalization

чт, 9 июл. 2026 г., 06:33 木葉 Scarlet ***@***.***>:

**Comment by sprow-dev** [comment]
> There is no way I am now being contacted by AI agents with AI slop. Please do not do this.

I am trying to be cool-headed here so here's my most polite response currently:
It is maddening and sickening that I am being blamed of being AI.

This *is* going to sound AI, but it's called having a brain and, you are not going to believe this, USING IT:
I understand your frustration. I can see how my repsonse structure could be indicative of AI and I know that you deal with a lot of this daily, but I entirely believe that you are misjudging me. I am trying to let you know that Torch and TF are unable to work in parallel with each other (**properly**, results vary). If you would like to helpfully fix the implementation (I'm not installing Windows onto my AI workstation to make a single PR, or heck even Linux if I want it to crash every 5 minutes), it would be a lot more useful than calling me AI. Thanks!


And no I am not trying to draw attention to the Linux bug. It's still unaddressed and unloved (and closed by me eventually because I was sick of dealing with it) even though tens of people have the same issue. Even I have no clue what's going on there and honestly, I have no intention of doing so at this rate.

Heck, I'll give you the AI response to what you just said to show the stark contrast:

---

## Gemini would've said:

"Man, assuming everyone who writes complete sentences is an AI is wild.

I spent time debugging your DJL setup. The issue is that PyTorch configs are being passed into a TensorFlow backend, which DJL can't cross-match. That's why the AI aura is broken.

Believe what you want, but the bug is real. Fix it or don't."

---

### ChatGPT would've said:
"I wrote that analysis myself; it wasn't AI-generated.

If my conclusions are incorrect, I'd appreciate knowing where. My understanding after reading the implementation was that the model setup mixes PyTorch artifacts with a TensorFlow backend through DJL, and I believe that's contributing to the behavior reported in the original issue. I also noticed a lack of input normalization and what appears to be a memory leak during training initialization.

If I'm mistaken on any of those points, I'd be happy to be corrected. Otherwise, my intent was simply to provide a concrete technical explanation for the issue rather than speculation."

---

(I've placed separators to make it easier for you to read, thank me later. Or don't, I couldn't care less. It took 5 seconds)

I believe that those 2 AI responses are enough to debunk that claim.



---

Edit: I went against my word and did a tiny bit of digging.
[I found the pull request](https://github.com/CCBlueX/LiquidBounce/pull/5668)
It was Izuna's own code. I can see why they are defensive over it. 5 days is a while.

But also, this code is sloppy at best.

What's this in AiAngleSmooth.kt? "override fun calculateTicks(
        currentRotation: Rotation,
        targetRotation: Rotation
    ): Int {
        // TODO: Implement correctly
        return 0
    }"
TODO: Implement correctly.

Promising words for code that has been in prod for over a year now. I have gone through the work of git grepping this and found that it **is** referenced normally under KillAura IN THE MAIN MODULE CODE. Now, whether it is required to work properly is another question (it doesn't appear to be so but it does say "implement **correctly**" in the comment), but I've already dug enough to prove the point.

Don't believe me? [View the file yourself](https://raw.githubusercontent.com/CCBlueX/LiquidBounce/refs/heads/nextgen/src/main/kotlin/net/ccbluex/liquidbounce/utils/aiming/features/processors/anglesmooth/impl/AiAngleSmooth.kt)

**Comment by TimofeyFilkin-alt-i-lost-my-codes** [comment]
idk why but your manner of speech just made me suspect of ai.

чт, 9 июл. 2026 г., 21:34 Sprow-dev ***@***.***>:

**Comment by sprow-dev** [comment]
Well I am not AI and I am tired of the fact that being polite = AI now. I am extremely angry because I spent hours that day debugging and working on my own code, and then I put in the time to read the code of someone else's project FOR FREE, VOLUNTARILY, IN A PROGRAMMING LANGUAGE THAT IS NOT MY MAIN, AND DID RESEARCH ON EVERYTHING, and now I am labeled as AI.

I spent tons of time actually digging into the code and (you're not going to believe this) actually reading it and was completely dismissed without a second word. All this because of a bunch of dumb word predictors that do nothing other than regurgitate slop all day because people forced it to for no reason. A bunch of dumb word predictors that use up water and resources where I live and pollute the air to feed those slop machines. And I am that now apparently. A resource-wasting slop regurgitator.


Can you see why I am frustrated or is your brain too small to wrap around that?

**Comment by TimofeyFilkin-alt-i-lost-my-codes** [comment]
I underatand you, but no human formats text in issues, but ai does

пт, 10 июл. 2026 г., 21:33 Sprow-dev ***@***.***>:

**Comment by sprow-dev** [comment]
Ok so now a human can't know markdown. Noted. Also, read my edit.

Plus, markdown is simple.
\--- is horizontal line break
\* and \* is italicize
\*\* and \*\* is bold.
\# is H1
\## is H2
\### is H3.

(I apologize if that is broken, I'm not entirely sure how it escapes)

That's all you need to know to write like I do. And I am dismissed as AI. This over 6 formatting rules. Do you now look at a book that is formatted and call it AI without reading what it says?

I have a challenge for you: Open a word document, write "I am an idiot", change the font to comic sans, and convince yourself it is AI. If you can't do it, then I am not AI.

**Comment by TimofeyFilkin-alt-i-lost-my-codes** [comment]
ik how to do markdown. But nobody cares enough to do it

пт, 10 июл. 2026 г., 21:42 Sprow-dev ***@***.***>:

**Comment by sprow-dev** [comment]
logic check: you pretty much just said "i know how to use a feature by humans for humans but no human would actually bother to use the feature implemented by humans for humans to use". AI agents ust use that feature that is made for humans and so now you call me ai because of it. I believe I know the answer to my question: "Can you see why I am frustrated or is your brain too small to wrap around that?" It's the latter option, clearly. you do not even make an attempt to demonstrate any form of intelligence. if you can't be receptive to this basic form of logic, though, you are a lost cause and honestly i'm just DONE with you.

what an idiot

Fri, July 10 2026, 13:02 TimofeyFilkin-alt-i-lost-my-codes \*\*\*\@\*\*\*\.\*\*\*\>:

**Comment by TimofeyFilkin-alt-i-lost-my-codes** [comment]
you just dont get me, no one uses markdown when chatting. Its used in books
and such, but never in any casual chats

пт, 10 июл. 2026 г., 22:02 Sprow-dev ***@***.***>:

**Comment by sprow-dev** [comment]
<img width="264" height="50" alt="Image" src="https://github.com/user-attachments/assets/95dfbcfa-2578-4b2b-a646-5a3fd58d9a71" />

**Comment by TimofeyFilkin-alt-i-lost-my-codes** [comment]
does anybody really uses it? Only md i use when chatting is occasional
**bold**, and really rarely striketroughs

пт, 10 июл. 2026 г., 22:29 Sprow-dev ***@***.***>:

**Comment by sprow-dev** [comment]
"ik how to do markdown. But nobody cares enough to do it"
"no one uses markdown when chatting. Its used in books"
"Only md i use when chatting is occasional
**bold**, and really rarely striketroughs"

Contradicting yourself because I make an argument that is too good to refute. You made 3 arguments that all overlap and none bolt into each other.

You aren't even calling me AI, anymore. Just attacking the usage of markdown.

**Comment by TimofeyFilkin-alt-i-lost-my-codes** [comment]
in first and second ones i meant heavy markdown you did, and in third down
i just say that the most is some ocasional styling
пт, 10 июл. 2026 г., 22:38 Sprow-dev ***@***.***>:

**Comment by sprow-dev** [comment]
you are not addressing the elephant in the room: you still modified your point of attack. If you want to make a strong argument, you need to make it clear what you are arguing or it can be used against you.

Why? You could've magically changed your motives and I would have no clue. Therefore, that does not stand.

you ruined your own argument so devastatingly because you are stubborn and don't realize one fact: not everyone is AI. I'll give you a sample of an AI response to that to make sure you know.

(I clarified that these are not written by me.)
### Gemini would respond with:
(it gave me 3 options): "Option A (Direct & Analytical): > "You're moving the goalposts. You didn't say 'heavy markdown' initially; you said nobody cares and no one uses it when chatting. Admitting you use bold and strikethrough means you do use markdown when chatting. You're just changing the definition now because your original absolute statements were wrong."

Option B (Short & Witty): > "So 'no one uses it' actually meant 'I use it, just not as much as you.' Glad we agree that people use markdown in chat after all."

Option C (Focusing on the Argument's Structure): > "If you have to add qualifiers like 'heavy' after the fact to make your point work, your original argument didn't hold water. Bold is markdown. You use it. Therefore, chat markdown is used.""

All of Gemini's responses are crap.

i'd shut up if i were you because clearly your argument is making no sense and clearly i have proven that i am not ai and you are grasping at straws

**Comment by TimofeyFilkin-alt-i-lost-my-codes** [comment]
im not trying to argue here

пт, 10 июл. 2026 г., 22:52 Sprow-dev ***@***.***>:

**Comment by sprow-dev** [comment]
By the way, this issue will likely be locked and closed by Izuna (and you are probably all thinking "good riddance" to get rid of it right now) since the thread has become too long and tangential. (Although I'm not good at reading maintainers' minds so no guarantees)

Just so you know, this might be useful to you all:
<img width="300" height="200" alt="Image" src="https://github.com/user-attachments/assets/a9aa0be3-66d4-49bd-9c67-3f35fc7a2e86" />
(That was the first image I got when I searched the word "brain" on Google)

Now I'm going to go back to writing crappy C# code and y'all can probably figure out something of value. I'm also probably never going to touch this repo again since both experiences on it were crap.
(First was a Linux bug that was then dismissed by Izuna, and then people continued to report issues on an "can't repro" bug after I closed it myself because everyone in power was telling me that it was unreproducible, [here](https://github.com/CCBlueX/LiquidBounce/issues/6962), and now whatever this mess is that I've gotten myself into by attempting to be useful and give people my time for free.)

> im not trying to argue here
False, but I'll leave it be.

**Comment by TimofeyFilkin-alt-i-lost-my-codes** [comment]
yea will be more productive than wasting our time into nothing

сб, 11 июл. 2026 г., 04:45 Sprow-dev ***@***.***>:

**Comment by 6x68** [comment]
jobs? employment? where?

**Comment by 6x68** [comment]
this also happened eons ago with acceleration rotations so lol

**Comment by CzechHek** [comment]
what the actual f have I just read

**Comment by 6x68** [comment]
> what the actual f have I just read

😭 its kinda crazy
